config/config.go

```diff
@@ -0,0 +1,213 @@
+	}
+
+	return errors.Join(errs...)
+}
+
+// validateUnrecognizedParameters validates that the config only contains
+// parameters specified in the parameters.
+func (c Config) validateUnrecognizedParameters(params Parameters) []error {
+	var errs []error
+	for key := range c {
+		if _, ok := params[key]; !ok {
+			errs = append(errs, fmt.Errorf("%q: %w", key, ErrUnrecognizedParameter))
+		}
+	}
+	return errs
+}
+
+// validateParamType validates that a parameter value is parsable to its assigned type.
+func (c Config) validateParamType(key string, param Parameter) error {
+	value := c[key]
+	// empty value is valid for all types
+	if c[key] == "" {
+		return nil
+	}
+
+	//nolint:exhaustive // type ParameterTypeFile and ParameterTypeString don't need type validations (both are strings or byte slices)
+	switch param.Type {
+	case ParameterTypeInt:
+		_, err := strconv.Atoi(value)
+		if err != nil {
+			return fmt.Errorf("error validating %q: %q value is not an integer: %w", key, value, ErrInvalidParameterType)
+		}
+	case ParameterTypeFloat:
+		_, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return fmt.Errorf("error validating %q: %q value is not a float: %w", key, value, ErrInvalidParameterType)
+		}
+	case ParameterTypeDuration:
+		_, err := time.ParseDuration(value)
+		if err != nil {
+			return fmt.Errorf("error validating %q: %q value is not a duration: %w", key, value, ErrInvalidParameterType)
+		}
+	case ParameterTypeBool:
+		_, err := strconv.ParseBool(value)
+		if err != nil {
+			return fmt.Errorf("error validating %q: %q value is not a boolean: %w", key, value, ErrInvalidParameterType)
+		}
+	}
+	return nil
+}
+
+// validateParamValue validates that a configuration value matches all the
+// validations required for the parameter.
+func (c Config) validateParamValue(key string, param Parameter) error {
+	value := c[key]
+	var errs []error
+
+	isRequired := false
+	for _, v := range param.Validations {
+		if _, ok := v.(ValidationRequired); ok {
+			isRequired = true
+		}
+		err := v.Validate(value)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error validating %q: %w", key, err))
+			continue
+		}
+	}
+	if value == "" && !isRequired {
+		return nil // empty optional parameter is valid
+	}
+
+	return errors.Join(errs...)
+}
+
+// DecodeInto copies configuration values into the target object.
+// Under the hood, this function uses github.com/mitchellh/mapstructure, with
+// the "mapstructure" tag renamed to "json". To rename a key, use the "json"
+// tag. To embed structs, append ",squash" to your tag. For more details and
+// docs, see https://pkg.go.dev/github.com/mitchellh/mapstructure.
+func (c Config) DecodeInto(target any) error {
+	dConfig := &mapstructure.DecoderConfig{
+		WeaklyTypedInput: true,
+		Result:           &target,
+		DecodeHook: mapstructure.ComposeDecodeHookFunc(
+			emptyStringToZeroValueHookFunc(),
+			mapstructure.StringToTimeDurationHookFunc(),
+			mapstructure.StringToSliceHookFunc(","),
```

Comment:
I've used a decode hook for field references, maybe it's good to add it too? We use it quite frequently.

Reply:
What exactly do you mean? 🤔

Reply:
I mean the SDK field reference resolvers, e.g. https://github.com/ConduitIO/conduit/pull/1382/files#diff-653885ccac432a28228f636af73cb2738506021f40693bc691cb0b369550725cR73

Reply:
Interesting! The issue is that the resolver lives in the processor SDK, we'd have to then bring that also into commons x) What if we expose a way to add additional hooks in this method? 🤔

Reply:
Yeah, I don't think we have a use case for references outside of processors.  I added a way to specify custom hooks here 👍

Reply:
Sweet!